PyInstaller/building/imphook.py

```diff
@@ -12,18 +12,573 @@
+        non-magic, and undefined attributes.
+
+        See Also
+        ----------
+        Class docstring for supported magic attributes.
+        '''
+
+        # If this is a magic attribute, initialize this attribute by lazy
+        # loading this hook script and then remove this attribute.
+        if attr_name in _MAGIC_MODULE_HOOK_ATTR_NAMES:
+            self._load_hook_module()
+            del self.__hook_attrs[attr_name]
+        # Else, this is a non-magic attribute. Defer to our superclass.
+        else:
+            return super(ModuleHook, self).__delattr__(attr_name)
+
+
+    ## Loading
+
+    def _load_hook_module(self):
+        """
+        Lazily load this hook script into an in-memory private module.
+
+        This method (and, indeed, this class) preserves all attributes and
+        functions defined by this hook script as is, ensuring sane behaviour in
+        hook functions _not_ expecting unplanned external modification. Instead,
+        this method copies public attributes defined by this hook script
+        (e.g., `binaries`) into private attributes of this object, which the
+        special `__getattr__()` and `__setattr__()` methods safely expose to
+        external callers. For public attributes _not_ defined by this hook
+        script, the corresponding private attributes will be assigned sane
+        defaults. For some public attributes defined by this hook script, the
+        corresponding private attributes will be transformed into objects more
+        readily and safely consumed elsewhere by external callers.
+
+        See Also
+        ----------
+        Class docstring for supported attributes.
+        """
+
+        # If this hook script module has already been loaded, noop.
+        if self._hook_module is not None:
+            return
+
+        # Load this hook script.
+        logger.info(
+            'Loading module hook "%s"...', os.path.basename(self.hook_filename))
+        self._hook_module = importlib_load_source(
+            self.hook_module_name, self.hook_filename)
+
+        # Copy attributes defined by this hook module into magic attributes
+        # exposed by this "ModuleHook" instance. Undefined attributes will
+        # retain their magic attribute defaults set by the __init__() method.
+        for attr_name, attr_sanitizer in (
+            # Convert hook-style 2-tuple lists into TOC-style 2-tuple sets.
+            # This must be done *BEFORE* the subsequent _process_hook_func()
+            # method adds already converted TOC-style 2-tuples to these sets.
+            ('binaries', format_binaries_and_datas),
+            ('datas',    format_binaries_and_datas),
+
+            # Convert lists in which order is insignificant into sets.
+            ('excludedimports', set),
+
+            # Preserve lists in which order is significant as lists.
+            ('hiddenimports', list),   # order of importation is significant
```

Comment:
**Hooks, mostly.**

The mere act of loading a hook script or calling a hook function defined by a hook script can have side effects conditionally dependent on module importation order. Since module importation triggers both the loading of hook scripts _and_ calling of hook functions defined by hook scripts, it follows that module importation itself can have these sorts of side effects.

_"Say wat-wat?_", you are thinking. I once thought like you. Now, consider two worst-case hook scripts `hook-Yog.Sothoth.py` and `hook-Shub.Niggurath.py` defined as follows:

* In `hook-Yog.Sothoth.py`:

        # I am evil hook script! I do what should not be done.
        from PyInstaller.depend import analysis
        if getattr(analysis, '_IS_GREAT_OLD_ONE_IMPORTED', False):
            hiddenimports = ['Azathoth', 'Ghatanothoa']
        else:
            analysis._IS_GREAT_OLD_ONE_IMPORTED = True

* In `hook-Shub.Niggurath.py`:

        # I am evil hook script also! I too do what should not be done.
        from PyInstaller.depend import analysis
        if getattr(analysis, '_IS_GREAT_OLD_ONE_IMPORTED', False):
            excludedimports = ['Nyarlathotep', 'Yig']
        else:
            analysis._IS_GREAT_OLD_ONE_IMPORTED = True

That's an extreme example. A saner example would probably have these two hook scripts define post-graph `hook()` functions conditionally adding hidden and/or excluded imports conditionally depending on whether or not certain modules had already been imported (e.g., by querying the current module graph).

In any case, order of importation is significant. Hooks are interpreted by the current Python process and hence have the keys to the PyInstaller kingdom. Although pathological hooks cannot be prevented from behaving pathologically, PyInstaller can at least ensure that it does its best to "do the right thing."

Of course, no well-behaved hook script should have these sorts of side effects. In the ideal world, _all_ hooks would refrain from modifying anything outside their jurisdiction (i.e., the local script in which they're defined) and would _not_ conditionally depend on module importation order. In the actual horrible world, hooks can anarchistically do anything they like – including monkey patch low-level PyInstaller and even lower-level Python operation.

Which means that somewhere, sometime, someone will do it. And they'll be really smug about it. And, when it burns down their multi-million dollar NSA datacenter, they'll post an accusatory PyInstaller issue. _And it's somehow all our fault._